src/pool/pool_jemalloc.c

```diff
@@ -0,0 +1,345 @@
+    if (new_ptr == NULL) {
+        TLS_last_allocation_error = UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+        return NULL;
+    }
+
+    return new_ptr;
+}
+
+static void *je_aligned_alloc(void *pool, size_t size, size_t alignment) {
+    jemalloc_memory_pool_t *je_pool = (jemalloc_memory_pool_t *)pool;
+    unsigned arena = je_pool->arena_zero;
+    int flags =
+        MALLOCX_ALIGN(alignment) | MALLOCX_ARENA(arena) | MALLOCX_TCACHE_NONE;
+    void *ptr = mallocx(size, flags);
+    if (ptr == NULL) {
+        TLS_last_allocation_error = UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+        return NULL;
+    }
+
+    return ptr;
+}
+
+static enum umf_result_t je_initialize(umf_memory_provider_handle_t hProvider,
+                                       void *params, void **out_pool) {
+    (void)params; // unused
+
+    extent_hooks_t *pHooks = &arena_extent_hooks;
+    size_t unsigned_size = sizeof(unsigned);
+    int err;
+
+    jemalloc_memory_pool_t *pool = malloc(sizeof(jemalloc_memory_pool_t));
+    if (!pool) {
+        return UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+    }
+
+    pool->arena_map_len = 1;
+    pool->hProvider = hProvider;
+
+    for (unsigned i = 0; i < pool->arena_map_len; i++) {
+        unsigned arena_index;
+        err = mallctl("arenas.create", (void *)&arena_index, &unsigned_size,
+                      NULL, 0);
+        if (err) {
+            fprintf(stderr, "Could not create arena.\n");
+            goto err_free_pool;
+        }
+
+        // setup extent_hooks for newly created arena
+        char cmd[64];
+        snprintf(cmd, sizeof(cmd), "arena.%u.extent_hooks", arena_index);
+        err =
+            mallctl(cmd, NULL, NULL, (void *)&pHooks, sizeof(extent_hooks_t *));
+        if (err) {
+            fprintf(stderr,
+                    "Could not setup extent_hooks for newly created arena.\n");
+            goto err_free_pool;
+        }
+
+        // store arena with lowest index (arenas could be created in
+        // descending/ascending order)
+        if (i == 0 || pool->arena_zero > arena_index) {
+            pool->arena_zero = arena_index;
+        }
+        arena_registry_g[arena_index] = pool;
+    }
+
+    *out_pool = (umf_memory_pool_handle_t)pool;
+
+    return UMF_RESULT_SUCCESS;
+
+err_free_pool:
+    free(pool);
+    return UMF_RESULT_ERROR_MEMORY_PROVIDER_SPECIFIC;
+}
+
+static void je_finalize(void *pool) {
+    jemalloc_memory_pool_t *je_pool = (jemalloc_memory_pool_t *)pool;
+    if (je_pool->arena_map_len) {
+        char cmd[128];
+        for (unsigned i = 0; i < je_pool->arena_map_len; ++i) {
+            snprintf(cmd, 128, "arena.%u.destroy", je_pool->arena_zero + i);
+            mallctl(cmd, NULL, NULL, NULL, 0);
+            arena_registry_g[je_pool->arena_zero + i] = NULL;
+        }
+    }
+    free(je_pool);
+}
+
+static size_t je_malloc_usable_size(void *pool, void *ptr) {
+    // not supported
+    (void)pool; // not used
+    (void)ptr;  // not used
+    return 0;
```

Comment:
please set here TLS_last_allocation_error. BTW - why we can't support this? Improve the comment.

Reply:
OK so at least print an error here

Reply:
Changed